File: ticketProject5-DistributedC/TicketSystemCC/persistance/Databases/ARDatabase.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using persistance.Interfaces;
using model;
using model.Validators;

namespace persistance.Database
{
    public class ARDatabase : IARRepository
    {
        //private static readonly ILog log = LogManager.GetLogger("ARDatabase");
        private Validator<ArtisticRepresentation> validator;

        public ARDatabase(Validator<ArtisticRepresentation> validator)
        {
            //log.Info("Creating ARDatabase");
            this.validator = validator;
        }

        public ArtisticRepresentation extractEntity(List<String> attributes)
        {
            if (attributes.Count < 6)
                throw new ValidationException("Insuficient datas!\n");
            ArtisticRepresentation ar = new ArtisticRepresentation(attributes[1], DateTime.Parse(attributes[2]), attributes[3], long.Parse(attributes[4]), long.Parse(attributes[5]))
            {
                Id = Int64.Parse(attributes[0])
            };
            return ar;
        }

        protected String createEntityAsString(ArtisticRepresentation entity)
        {
            return entity.Id + ";" + entity.ArtistName + ";" + entity.Data + ";" + entity.Location + ";" + entity.AvailableSeats + ";" + entity.SoldSeats;
        }

        public ArtisticRepresentation FindOne(long id)
        {
            IDbConnection con = DBUtils.getConnection();

            try
            {
                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "select id, artistName, data, location, availableSeats, soldSeats from ARs where id=@id";
                    IDbDataParameter paramId = comm.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    comm.Parameters.Add(paramId);

                    using (var dataR = comm.ExecuteReader())
                    {
                        if (dataR.Read())
                        {
                            long idd = dataR.GetInt32(0);
                            String artistName = dataR.GetString(1);
                            DateTime data = DateTime.Parse(dataR.GetString(2));
                            String location = dataR.GetString(3);
                            long availableSeats = dataR.GetInt64(4);
                            long soldSeats = dataR.GetInt64(5);

                            ArtisticRepresentation ar = new ArtisticRepresentation(artistName, data, location, availableSeats, soldSeats)
                            {
                                Id = idd
                            };

                            Console.WriteLine(ar);
                            return ar;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Mesajul erorii este {0}", ex.Message);
            }
            return null;
        }

        public IEnumerable<ArtisticRepresentation> FindAll()
        {
            Console.WriteLine("sunt in ARrepo");
            IDbConnection con = DBUtils.getConnection();
            Console.WriteLine("sunt in ARrepo2");
            IList<ArtisticRepresentation> ars = new List<ArtisticRepresentation>();

            try
            {
                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "select id, artistName, data, location, availableSeats, soldSeats from ARs";

                    using (var dataR = comm.ExecuteReader())
                    {
                        while (dataR.Read())
                        {
                            long idd = dataR.GetInt32(0);
                            String artistName = dataR.GetString(1);
                            DateTime data = DateTime.Parse(dataR.GetString(2));
                            String location = dataR.GetString(3);
                            long availableSeats = dataR.GetInt64(4);
                            long soldSeats = dataR.GetInt64(5);

                            ArtisticRepresentation ar = new ArtisticRepresentation(artistName, data, location, availableSeats, soldSeats)
                            {
                                Id = idd
                            };

                            ars.Add(ar);
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Mesajul erorii este {0}", ex.Message);
            }

            return ars;
        }


        public ArtisticRepresentation Save(ArtisticRepresentation entity)
        {
            if (entity == null)
                throw new ArgumentException("entity must be not null");

            validator.Validate(entity);
            //if (FindOne(entity.Id) != null)
            //{
            //    return entity;
            //}

            var con = DBUtils.getConnection();
            try
            {
                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "insert into ARs (artistName, data, location, availableSeats, soldSeats)  values (@artistName, @data, @location, @availableSeats, @soldSeats)";

                    var paramArtistName = comm.CreateParameter();
                    paramArtistName.ParameterName = "@artistName";
                    paramArtistName.Value = entity.ArtistName;
                    comm.Parameters.Add(paramArtistName);

                    var paramData = comm.CreateParameter();
                    paramData.ParameterName = "@data";
                    paramData.Value = entity.Data;
                    comm.Parameters.Add(paramData);

                    var paramLocation = comm.CreateParameter();
                    paramLocation.ParameterName = "@location";
                    paramLocation.Value = entity.Location;
                    comm.Parameters.Add(paramLocation);

                    var paramAvailableSeats = comm.CreateParameter();
                    paramAvailableSeats.ParameterName = "@availableSeats";
                    paramAvailableSeats.Value = entity.AvailableSeats;
                    comm.Parameters.Add(paramAvailableSeats);

                    var paramSoldSeats = comm.CreateParameter();
                    paramSoldSeats.ParameterName = "@soldSeats";
                    paramSoldSeats.Value = entity.SoldSeats;
                    comm.Parameters.Add(paramSoldSeats);

                    var result = comm.ExecuteNonQuery();
                    if (result == 0)
                        throw new Exception("Insert failed !");
                    //selse
                    //log.InfoFormat("Saved the AR {0}", entity);
                }
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Mesajul erorii este {0}", ex.Message);
            }

            return null;
        }


        public ArtisticRepresentation Delete(long id)
        {
            if (id == null)
                throw new ArgumentException("Non existent user!");

            var con = DBUtils.getConnection();
            try
            {
                ArtisticRepresentation entity = FindOne(id);

                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "delete from ARs where id=@id";
                    IDbDataParameter paramId = comm.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = id;
                    comm.Parameters.Add(paramId);
                    var result = comm.ExecuteNonQuery();
                    if (result == 0)
                        return null;
                    else
                    {
                        //log.InfoFormat("Deleted AR {0}", entity);
                        return entity;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Mesajul erorii este {0}", ex.Message);
            }
            return null;
        }

        public ArtisticRepresentation Update(ArtisticRepresentation entity)
        {
            Debug.WriteLine("2 " + entity.SoldSeats);
            if (entity == null)
                throw new ArgumentException("entity must be not null!");
            validator.Validate(entity);

            if (FindOne(entity.Id) == null)
                return entity;

            var con = DBUtils.getConnection();
            try
            {
                using (var comm = con.CreateCommand())
                {
                    comm.CommandText = "update ARs set artistName=(@artistName), data=(@data), location=(@location), availableSeats=(@availableSeats), soldSeats=(@soldSeats) where id=(@id)";

                    var paramArtistName = comm.CreateParameter();
                    paramArtistName.ParameterName = "@artistName";
                    paramArtistName.Value = entity.ArtistName;
                    comm.Parameters.Add(paramArtistName);

                    var paramData = comm.CreateParameter();
                    paramData.ParameterName = "@data";
                    paramData.Value = entity.Data;
                    comm.Parameters.Add(paramData);

                    var paramLocation = comm.CreateParameter();
                    paramLocation.ParameterName = "@location";
                    paramLocation.Value = entity.Location;
                    comm.Parameters.Add(paramLocation);

                    var paramAvailableSeats = comm.CreateParameter();
                    paramAvailableSeats.ParameterName = "@availableSeats";
                    paramAvailableSeats.Value = entity.AvailableSeats;
                    comm.Parameters.Add(paramAvailableSeats);

                    var paramSoldSeats = comm.CreateParameter();
                    paramSoldSeats.ParameterName = "@soldSeats";
                    paramSoldSeats.Value = entity.SoldSeats;
                    comm.Parameters.Add(paramSoldSeats);
                    //Debug.WriteLine("3 " + paramSoldSeats.Value);

                    var paramId = comm.CreateParameter();
                    paramId.ParameterName = "@id";
                    paramId.Value = entity.Id;
                    comm.Parameters.Add(paramId);

                    var result = comm.ExecuteNonQuery();
                    if (result == 0)
                        return entity;
                    else
                    {
                        //log.InfoFormat("Updated AR {0}", entity);
                        return null;
                    }
                }
            }
            catch (Exception ex)
            {
                Console.ForegroundColor = ConsoleColor.Red;
                Console.WriteLine("Mesajul erorii este {0}", ex.Message);
            }
            return entity;
        }
    }
}

File: ticketProject7-CrossPlatform/TS-cross/TicketPersistance/src/main/java/ticket/persistance/jdbc/ARDatabase.java
package ticket.persistance.jdbc;

import ticket.model.ArtisticRepresentation;
import ticket.model.validators.ValidationException;
import ticket.model.validators.Validator;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import ticket.persistance.ARRepository;
import ticket.persistance.jdbc.utils.JdbcUtils;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ARDatabase implements ARRepository {

    private final JdbcUtils dbUtils;
    final private Validator<ArtisticRepresentation> validator;
    private static final Logger logger= LogManager.getLogger();

    public ARDatabase(Validator<ArtisticRepresentation> validator, Properties props) {
        this.validator = validator;
        dbUtils=new JdbcUtils(props);
        logger.info("Initializing ARDatabase");
    }

    private LocalDateTime fromStringToLocalDateTime(String string){
        LocalDateTime data=LocalDateTime.now();
        try{
            data = LocalDateTime.parse(string);
        }
        catch (Exception ex){
            String[] arr = string.split(" ");
            String[] arrDate = arr[0].split("-");
            String[] arrTime = arr[1].split(":");

            data = LocalDateTime.of(Integer.parseInt(arrDate[0]),Integer.parseInt(arrDate[1]), Integer.parseInt(arrDate[2]), Integer.parseInt(arrTime[0]), Integer.parseInt(arrTime[1]));
        }
        return data;
    }

    public ArtisticRepresentation extractEntity(List<String> attributes) {
        if (attributes.size() < 6)
            throw new ValidationException("Insuficient datas!\n");
        ArtisticRepresentation ar = new ArtisticRepresentation(attributes.get(1), LocalDateTime.parse(attributes.get(2)), attributes.get(3), Long.parseLong(attributes.get(4)), Long.parseLong(attributes.get(5)));
        ar.setId(Long.parseLong(attributes.get(0)));
        return ar;
    }

    protected String createEntityAsString(ArtisticRepresentation entity) {
        return entity.getId() + ";" + entity.getArtistName() + ";" + entity.getData() + ";" + entity.getLocation() + ";" + entity.getAvailableSeats()+ ";"+entity.getSoldSeats();
    }


    @Override
    public ArtisticRepresentation findOne(Long id) {
        if (id == null)
            throw new IllegalArgumentException("id must be not null");

        Connection con= dbUtils.getConnection();
        try {
            PreparedStatement preStmt = con.prepareStatement("select * from ARs where id = ?");
            preStmt.setInt(1, Math.toIntExact(id));
            ResultSet result = preStmt.executeQuery();

            if (result.next()) {
                Long idd = result.getLong("id");
                String artistName = result.getString("artistName");
                String location = result.getString("location");
                LocalDateTime data = this.fromStringToLocalDateTime(result.getString("data"));
                Long availableSeats = result.getLong("availableSeats");
                Long soldSeats = result.getLong("soldSeats");

                ArtisticRepresentation ar = new ArtisticRepresentation(artistName, data, location, availableSeats, soldSeats);
                ar.setId(idd);

                result.close();
                preStmt.close();

                return ar;
            }
        } catch (Exception e) {
            System.out.println(e.toString());
        }
        return null;
    }

    public Iterable<ArtisticRepresentation>  findAllARWithARName(String name) {
        if (name == null)
            throw new IllegalArgumentException("name must be not null");

        Connection con= dbUtils.getConnection();
        List<ArtisticRepresentation> ars= new ArrayList<>();
        try {
            PreparedStatement preStmt = con.prepareStatement("select * from ARs where artistName = ?");
            preStmt.setString(1, name);
            ResultSet result = preStmt.executeQuery();
            while(result.next()){
                Long id=result.getLong("id");
                String artistName = result.getString("artistName");
                LocalDateTime data = this.fromStringToLocalDateTime(result.getString("data"));
                String location = result.getString("location");
                Long availableSeats = result.getLong("availableSeats");
                Long soldSeats = result.getLong("soldSeats");

                ArtisticRepresentation ar = new ArtisticRepresentation(artistName, data, location, availableSeats, soldSeats);
                ar.setId(id);

                ars.add(ar);
            }
            result.close();
            preStmt.close();
        } catch (Exception e) {
            System.out.println(e.toString());
        }
        return ars;
    }


    @Override
    public Iterable<ArtisticRepresentation> findAll(){
        Connection con= dbUtils.getConnection();
        List<ArtisticRepresentation> ars= new ArrayList<>();

        try{
            PreparedStatement preStmt = con.prepareStatement("select * from ARs ");
            ResultSet result = preStmt.executeQuery();
            while(result.next()){
                Long id=result.getLong("id");
                String artistName = result.getString("artistName");
                LocalDateTime data = this.fromStringToLocalDateTime(result.getString("data"));
                String location = result.getString("location");
                Long availableSeats = result.getLong("availableSeats");
                Long soldSeats = result.getLong("soldSeats");

                ArtisticRepresentation ar = new ArtisticRepresentation(artistName, data, location, availableSeats, soldSeats);
                ar.setId(id);

                ars.add(ar);
            }
            result.close();
            preStmt.close();
        }
        catch(SQLException ex){
            System.err.println("Error DB" + ex);
        }

        return ars;
    }


    @Override
    public ArtisticRepresentation save(ArtisticRepresentation entity) {
        logger.traceEntry("saving AR {}", entity);
        if (entity == null)
            throw new IllegalArgumentException("entity must be not null");
        validator.validate(entity);
//        if (findOne(entity.getId()) != null) {
//            return entity;
//        }

        Connection con= dbUtils.getConnection();
        try{
            PreparedStatement preStmt = con.prepareStatement("insert into ARs (artistName, data, location, availableSeats, soldSeats) values (?,?,?,?,?)");
            preStmt.setString(1,entity.getArtistName());
            //preStmt.setTimestamp(2, Timestamp.valueOf(entity.getData()));
            //preStmt.setDate(2, Date.valueOf(String.valueOf(entity.getData())));
            //preStmt.setDate(2, Date.valueOf(String.valueOf(entity.getData())));
            //preStmt.setTime(2, Time.valueOf(String.valueOf(entity.getData())));
            preStmt.setString(2, entity.getData().toString());
            preStmt.setString(3,entity.getLocation());
            preStmt.setInt(4, Math.toIntExact(entity.getAvailableSeats()));
            preStmt.setInt(5, Math.toIntExact(entity.getSoldSeats()));
            int result=preStmt.executeUpdate();

            preStmt.close();

            logger.trace("Saved {} instances", result);
            if (result == 0)
                throw new Exception("Insert Failed");
        }
        catch(SQLException ex){
            System.err.println("Error DB" + ex);
        }
        catch(Exception ex){
            logger.error(ex);
            System.err.println(ex);
        }

        logger.traceExit();
        return null;
    }


    @Override
    public ArtisticRepresentation delete(Long id) {
        logger.traceEntry("deleting AR with id {}", id);
        if (id == null)
            throw new IllegalArgumentException("Non existent user!");

        Connection con= dbUtils.getConnection();
        try {
            ArtisticRepresentation entity = findOne(id);

            PreparedStatement preStmt = con.prepareStatement("DELETE FROM ARs WHERE id = ?");
            preStmt.setInt(1, Math.toIntExact(id));
            int result = preStmt.executeUpdate();

            preStmt.close();

            //con.close();
            logger.trace("Deleted {} instances", result);
            if (result > 0)
                return entity;
            else
                return null;

        } catch (Exception e) {
            logger.error(e);
            System.out.println(e.toString());
        }

        logger.traceExit();
        return null;
    }


    @Override
    public ArtisticRepresentation update(ArtisticRepresentation entity) {
        logger.traceEntry("updating AR {}", entity);
        if (entity == null)
            throw new IllegalArgumentException("entity must be not null!");
        validator.validate(entity);

        if (findOne(entity.getId()) == null)
            return entity;
        Connection con= dbUtils.getConnection();
        try {
            PreparedStatement preStmt = con.prepareStatement("update ARs set artistName=(?), data=(?), location=(?), availableSeats=(?), soldSeats=(?) where id=(?);");
            String data = entity.getData().getYear()+"-"+entity.getData().getMonth()+"-"+entity.getData().getDayOfMonth()+" "+entity.getData().getHour()+":"+entity.getData().getMinute();
            preStmt.setString(1, entity.getArtistName());
            preStmt.setString(2, entity.getData().toString());
            //preStmt.setString(2, data);
            preStmt.setString(3, entity.getLocation());
            preStmt.setInt(4, Math.toIntExact(entity.getAvailableSeats()));
            preStmt.setInt(5, Math.toIntExact(entity.getSoldSeats()));
            preStmt.setInt(6, Math.toIntExact(entity.getId()));
            int result = preStmt.executeUpdate();

            preStmt.close();

            logger.trace("Updated {} instances", result);
            if (result > 0)
                return null;
            else
                return entity;
        } catch (Exception e) {
            logger.error(e);
            System.out.println(e.toString());
        }

        logger.traceExit();
        return entity;
    }


}
